### DISCUSSION TITLE
Use biome as custom formatter

### DISCUSSION THREAD

**Discussion by wcastand** [zed-industries/zed#6673] [Feature Requests]
i tried to use biome as formatter for my typescript project.

i have this config but it doesn't format on save nor when i use the command `editor:format`:

```
"format_on_save": "on",
  "formatter": {
    "external": {
      "command": "biome",
      "arguments": ["format", ".", "--write"]
    }
  },
```
 I tried to replace `.` by `{buffer_path}` since i was that in an issue but doesn't do anything either.
 
 I saw the ticket, support biome, does that mean we can't use biome as custom formatter and need a special integration?

**Comment by wcastand** [comment]
i managed to make it work altho it's not perfect.

first the best would be to use lsp since biome support it.

but in the meantime you can get it to work with this config:

```
  "formatter": {
    "external": {
      "command": "biome",
      "arguments": ["format", "--write", "--stdin-file-path={buffer_path}"]
    }
  },
```

the only issue is that you need to install biome as a global package, which means it doesn't use the local biome of your project and the potentially not using the right version is the global one is not in sync with local one.

is there anyway to target the command in the local node_modules

`./node_modules/.bin/biome` doesn't seems to work either.

you offer a `bufffer_path`, maybe you have a `workspace_path` or something so i can target the biome executable in my workspace?

**Reply by tebuevd** [reply]
+1 to this. We need biome to work on Zed :)

**Reply by lewisl9029** [reply]
Appreciate the clues! I did actually manage to get this working without installing biome globally by creating a local `.zed/settings.json` with the following contents:

```json
{
  "formatter": {
    "external": {
      "command": "./node_modules/.bin/biome",
      "arguments": [
        "format",
        "--write",
        "--stdin-file-path={buffer_path}"
      ]
    }
  }
}
```

This is documented here: https://docs.zed.dev/configuration/configuring-zed#folder-specific-settings

**Reply by byteoxo** [reply]
this is great, thank you!

**Comment by JosephTLyons** [comment]
Hey @wcastand, thanks for providing what you've found. We're keeping our eye on Biome - looks really great. Do you happen to know if they are planning on providing features similar to `tsserver`? Right now, I see that it lints and formats, curious if it will have things like go to def and autocompletions at some point.

**Reply by JosephTLyons** [reply]
Looks like I got my [answer](https://twitter.com/MyDeathMachine/status/1732861978199106006)!

**Reply by wcastand** [reply]
Can't answer for them but the goal at the start was on time to support all of the tool a dev need so replace tsc and stuff but I don't know if that still the case now. Would to check with maintainers of biome 

Sorry didn't see you had your answer already on mobile 😅 good then.

**Comment by enqqi** [comment]
There's an official zed extension by biome, it works but has some bugs

[biomejs/biome-zed](https://github.com/biomejs/biome-zed)

**Reply by firu11** [reply]
Doesn't do anything for me.

```
2025-08-27T20:50:53+02:00 INFO  [project.lsp_format] Formatting via LSP
2025-08-27T20:50:53+02:00 ERROR [project::lsp_store] Formatting failed: failed to format via language server: Biome doesn't support this feature for the language TypeScript
2025-08-27T20:50:53+02:00 ERROR [editor] failed to format via language server
```

**Comment by JosephTLyons** [comment]
The Zed extension store has a Biome extension now:

<img width="825" alt="SCR-20240528-ttyy" src="https://github.com/zed-industries/zed/assets/19867440/18c2ac50-0ed9-4270-9719-bf0f53f07539">

**Reply by unitythemaker** [reply]
Hey, I wonder if there is a way to both format and lint when I press Ctrl+S. From my understanding, we can only run one command like this:
```
    "external": {
      "command": "./node_modules/@biomejs/biome/bin/biome",
      "arguments": [
        "format",
        "--write",
        "--stdin-file-path",
        "{buffer_path}"
      ]
    }
```

**Reply by wottpal** [reply]
did you found a solution to this @unitythemaker? doesn't seem to be possible with the biome extension 🤔 might be worth an issue there

**Reply by unitythemaker** [reply]
I switched to eslint, unfortunately.

**Reply by wottpal** [reply]
It seems to work with the `check` command (just replace `format` with `check` in your snippet above). This should still be the default in the extension IMO. 

Also, as I didn't found a way to specify settings for multiple languages at once, the `.zed/settings.json` is becoming a small mess now like:

```jsonc
  "languages": {
    "TypeScript": {
      "formatter": {
        "external": {
          "command": "./node_modules/@biomejs/biome/bin/biome",
          "arguments": ["check", "--write", "--stdin-file-path", "{buffer_path}"]
        }
      }
    },
    "TSX": {
      "formatter": {
        "external": {
          "command": "./node_modules/@biomejs/biome/bin/biome",
          "arguments": ["check", "--write", "--stdin-file-path", "{buffer_path}"]
        }
      }
    },
    "JavaScript": {
      "formatter": {
        "external": {
          "command": "./node_modules/@biomejs/biome/bin/biome",
          "arguments": ["check", "--write", "--stdin-file-path", "{buffer_path}"]
        }
      }
    },
    "JSX": {
      "formatter": {
        "external": {
          "command": "./node_modules/@biomejs/biome/bin/biome",
          "arguments": ["check", "--write", "--stdin-file-path", "{buffer_path}"]
        }
      }
    },
    "JSON": {
      "formatter": {
        "external": {
          "command": "./node_modules/@biomejs/biome/bin/biome",
          "arguments": ["check", "--write", "--stdin-file-path", "{buffer_path}"]
        }
      }
    },

  // Defining prettier for all other languages that biome doesn't fully support yet
  // [...]
}
```

**Reply by wottpal** [reply]
cc @luckydye (sorry for tagging you twice, just think these are super relevant to make biome more usable in zed) 🫶

**Reply by luckydye** [reply]
With the latest versions of Zed and the biome extension, this is not needed anymore. The LSP can do the formatting now. [See](https://github.com/biomejs/biome-zed/blob/main/README.md) for correct configuration.

**Reply by wottpal** [reply]
Tried this just yesterday and didn't get it to apply lint fixes. Are you sure they are applied with those settings (I tested exactly those with latest versions of both). My setup above seems to work. 🤷‍♂️

**Reply by luckydye** [reply]
> Tried this just yesterday and didn't get it to apply lint fixes. Are you sure they are applied with those settings (I tested exactly those with latest versions of both). My setup above seems to work. 🤷‍♂️

Lint fixes are applied using code actions. Not formatting. If code actions don't work, it's a different issue.

**Reply by abue-ammar** [reply]
> It seems to work with the `check` command (just replace `format` with `check` in your snippet above). This should still be the default in the extension IMO.
> 
> Also, as I didn't found a way to specify settings for multiple languages at once, the `.zed/settings.json` is becoming a small mess now like:
> 
> ```js
>   "languages": {
>     "TypeScript": {
>       "formatter": {
>         "external": {
>           "command": "./node_modules/@biomejs/biome/bin/biome",
>           "arguments": ["check", "--write", "--stdin-file-path", "{buffer_path}"]
>         }
>       }
>     },
>     "TSX": {
>       "formatter": {
>         "external": {
>           "command": "./node_modules/@biomejs/biome/bin/biome",
>           "arguments": ["check", "--write", "--stdin-file-path", "{buffer_path}"]
>         }
>       }
>     },
>     "JavaScript": {
>       "formatter": {
>         "external": {
>           "command": "./node_modules/@biomejs/biome/bin/biome",
>           "arguments": ["check", "--write", "--stdin-file-path", "{buffer_path}"]
>         }
>       }
>     },
>     "JSX": {
>       "formatter": {
>         "external": {
>           "command": "./node_modules/@biomejs/biome/bin/biome",
>           "arguments": ["check", "--write", "--stdin-file-path", "{buffer_path}"]
>         }
>       }
>     },
>     "JSON": {
>       "formatter": {
>         "external": {
>           "command": "./node_modules/@biomejs/biome/bin/biome",
>           "arguments": ["check", "--write", "--stdin-file-path", "{buffer_path}"]
>         }
>       }
>     },

>   // Defining prettier for all other languages that biome doesn't fully support yet
>   // [...]
> }
> ```

well what if my another project which needs prettier. why it has to be from node modules? isn't there any proper way like global? don't tell me to install prettier globally with npm that's BS

**Reply by onx2** [reply]
FWIW, I installed the extension, followed these [instructions](https://biomejs.dev/reference/zed/), then double checked my global and local settings weren't conflicting - such an easy thing to miss. After doing that everything is working as expected :)
